### PR TITLE
Notifications

### PR DESCRIPTION
**Pull request by JimenaAndrea**
add notifications when a timer starts and when an activity with a timer is completed

### CODE REVIEW

File: App.js [hunk 1]
```diff
@@ -22,6 +23,16 @@ import { Drawer as CustomDrawer } from './src/components'
...
+Notifications.setNotificationHandler({
+  handleNotification: async () => {
+    return {
+      shouldShowAlert: true,
+      shouldPlaySound: true,
+      shouldSetBadge: true,
+    };
+  },
+});
+
```
Comment (OliverLSanz) [inline App.js:35]: What does this code do? Do notification alerts appear while the app is open if we set shouldShowAlert to false? Maybe this is the desired behavior.
```suggestion
Notifications.setNotificationHandler({
  handleNotification: async () => {
    return {
      shouldShowAlert: false,
      shouldPlaySound: false,
      shouldSetBadge: true,
    };
  },
});

```

Comment (OliverLSanz) [reply App.js:35]: Also, this code should be a function of the notifications module.

File: src/activityHandler/activityTypes/dailyGoals/doNSeconds.js [hunk 2]
```diff
@@ -31,20 +36,33 @@ const TodayScreenItem = ({ activityId, date }) => {
...
+ 
   // function definitions
   function onPressPause(){
+    //Stop timer
     if(date.toISO() == todayDate.toISO()){
       dispatch(stopTodayTimer( activityId ))
     }
+    //Dismiss notifications
+    Notifications.dismissNotificationAsync('timer')
+    Notifications.cancelScheduledNotificationAsync('complete' + activityId)
```
Comment (OliverLSanz) [inline src/activityHandler/activityTypes/dailyGoals/doNSeconds.js:52]: Move this logic to the notifications module

File: src/activityHandler/activityTypes/dailyGoals/doNSeconds.js [hunk 3]
```diff
@@ -31,20 +36,33 @@ const TodayScreenItem = ({ activityId, date }) => {
...
+    //Start Timer
     if(date.toISO() == todayDate.toISO()){
      dispatch(startTodayTimer( activityId ))
     }
+    //Send timer notification
+    timerNotification(activity, t)
+    //Schedule complete notification
+    if(!entry.completed){
+      completeScheduleNotification(activity, activityId, secondsRemaining, t)
+    }
```
Comment (OliverLSanz) [inline src/activityHandler/activityTypes/dailyGoals/doNSeconds.js:65]: Move this logic to a function in the notifications module

File: src/activityHandler/activityTypes/doNSecondsEachWeek.js [hunk 4]
```diff
@@ -37,18 +39,30 @@ const TodayScreenItem = ({ activityId, date }) => {
...
 
   // function definitions
   function onPressPause(){
+    //Stop Timer
     if(date.toISO() == todayDate.toISO()){
       dispatch(stopTodayTimer( activityId ))
     }
+    //Dismiss notifications
+    Notifications.dismissNotificationAsync('timer')
+    Notifications.cancelScheduledNotificationAsync('complete' + activityId)
```
Comment (OliverLSanz) [inline src/activityHandler/activityTypes/doNSecondsEachWeek.js:52]: Move this logic to a function in the notifications module

File: src/activityHandler/activityTypes/doNSecondsEachWeek.js [hunk 5]
```diff
@@ -37,18 +39,30 @@ const TodayScreenItem = ({ activityId, date }) => {
...
+    //Start Timer
     if(date.toISO() == todayDate.toISO()){
      dispatch(startTodayTimer( activityId ))
     }
+    //Send timer notification
+    timerNotification(activity, t)
+    //Schedule complete notification
+    if(!entry.completed){
+      completeScheduleNotification(activity, activityId, secondsRemaining, t)
+    }
```
Comment (OliverLSanz) [inline src/activityHandler/activityTypes/doNSecondsEachWeek.js:65]: Move this logic to a function in the notifications module

File: src/i18n/english.js [hunk 6]
```diff
@@ -388,6 +388,17 @@ const en = {
...
       appName: 'Goaliath'
+    },
+
+    notifications: {
+      timer: {
+        title: 'You are working on: ',
+        body: '{{activityName}}'
+      },
+      complete: {
+        title: 'Yoy have completed the activity: ',
```
Comment (OliverLSanz) [inline src/i18n/english.js:399]:
```suggestion
        title: 'You have completed the activity: ',
```

File: src/screens/ActivityDetailScreen/TodayPannel.js [hunk 7]
```diff
@@ -9,110 +10,127 @@ import { useTranslation } from 'react-i18next'
...
-    const { t, i18n } = useTranslation()
+  const { t, i18n } = useTranslation()
   
-    function onPressPlay(){
-      startTodayTimer(entry.id)
+  const [todayTime, setTodayTime] = React.useState(getTodayTime(entry.intervals))
+
+  const secondsGoal = activity.type=='doFixedDays'||activity.type=='doNDaysEachWeek'? activity.params.dailyGoal.params.seconds : 
+                      activity.type=='doNSecondsEachWeek'? activity.params.seconds : null
+  const secondsRemaining = secondsGoal - todayTime.as('seconds')
```
Comment (OliverLSanz) [inline src/screens/ActivityDetailScreen/TodayPannel.js:31]: Add to the activityHandler a function to retrieve the time goal of an activity, that returns null if the activity does not have a time goal. Substitute this code by a call to that function.

This way we can create new activity types without modifying this file.

File: src/screens/ActivityDetailScreen/TodayPannel.js [hunk 8]
```diff
@@ -9,110 +10,127 @@ import { useTranslation } from 'react-i18next'
...
-const TodayPannel = ({ entry, toggleCompleted, startTodayTimer, stopTodayTimer, upsertEntry, date, dayStartHour }) => {
-    React.useEffect(() => {
-      if (isActivityRunning(entry.intervals)) {
-        const intervalId = setInterval(() => {
-          setTodayTime(getTodayTime(entry.intervals))    
-        }, 1000)
-        return () => clearInterval(intervalId)
-      }
-    }, [entry.intervals])
+const TodayPannel = ({ entry, toggleCompleted, startTodayTimer, stopTodayTimer, upsertEntry, date, dayStartHour, activity }) => {
```
Comment (OliverLSanz) [inline src/screens/ActivityDetailScreen/TodayPannel.js:15]: After resolving all comments, ensure that you still need the activity as a prop.

File: src/screens/ActivityDetailScreen/TodayPannel.js [hunk 9]
```diff
@@ -9,110 +10,127 @@ import { useTranslation } from 'react-i18next'
...
+
+  function onPressPlay(){
+    //Start timer
+    startTodayTimer(entry.id)
+    //Send timer notification
+    timerNotification(activity, t)
+    //Schedule complete notification
+    if(!entry.completed){
+      completeScheduleNotification(activity, activity.id, secondsRemaining, t)
     }
```
Comment (OliverLSanz) [inline src/screens/ActivityDetailScreen/TodayPannel.js:41]: Move this logic to a function in the notifications module

File: src/screens/ActivityDetailScreen/TodayPannel.js [hunk 10]
```diff
@@ -9,110 +10,127 @@ import { useTranslation } from 'react-i18next'
...
+  }
+
+  function onPressPause(){
+    //Stop timer
+    stopTodayTimer(entry.id)
+    //Dismiss notifications
+    Notifications.dismissNotificationAsync('timer')
+    Notifications.cancelScheduledNotificationAsync('complete' + activity.id)
```
Comment (OliverLSanz) [inline src/screens/ActivityDetailScreen/TodayPannel.js:49]: Move this logic to a function in the notifications module

File: src/screens/ActivityDetailScreen/ActivityDetailScreen.js [hunk 11]
```diff
@@ -128,7 +128,8 @@ const ActivityDetailScreen = ({
           stopTodayTimer={stopTodayTimer} 
           upsertEntry={upsertEntry} 
           date={date} 
-          dayStartHour={dayStartHour} 
+          dayStartHour={dayStartHour}
+          activity={activity}
```
Comment (OliverLSanz) [inline src/screens/ActivityDetailScreen/ActivityDetailScreen.js:132]: After resolving the rest of comments, ensure that this prop is still needed.

**Review by OliverLSanz — CHANGES_REQUESTED** [review]

File: src/i18n/english.js (hunk 6 above)
Comment (JimenaAndrea) [reply src/i18n/english.js]: Fixed by d5d71f30c649664682eef3584660c01c65376dc7

File: src/screens/ActivityDetailScreen/ActivityDetailScreen.js (hunk 11 above)
Comment (JimenaAndrea) [reply src/screens/ActivityDetailScreen/ActivityDetailScreen.js:132]: Yes, the prop is still needed

File: src/screens/ActivityDetailScreen/TodayPannel.js (hunk 10 above)
Comment (JimenaAndrea) [reply src/screens/ActivityDetailScreen/TodayPannel.js]: Moved in da9cadbaf09044c09bd762d03fb08d007a64ea5a

File: src/screens/ActivityDetailScreen/TodayPannel.js (hunk 9 above)
Comment (JimenaAndrea) [reply src/screens/ActivityDetailScreen/TodayPannel.js]: Moved in da9cadbaf09044c09bd762d03fb08d007a64ea5a

File: src/screens/ActivityDetailScreen/TodayPannel.js (hunk 8 above)
Comment (JimenaAndrea) [reply src/screens/ActivityDetailScreen/TodayPannel.js:14]: Yes, the prop is still needed

File: src/activityHandler/activityTypes/doNSecondsEachWeek.js (hunk 5 above)
Comment (JimenaAndrea) [reply src/activityHandler/activityTypes/doNSecondsEachWeek.js]: Moved in da9cadbaf09044c09bd762d03fb08d007a64ea5a

File: src/activityHandler/activityTypes/doNSecondsEachWeek.js (hunk 4 above)
Comment (JimenaAndrea) [reply src/activityHandler/activityTypes/doNSecondsEachWeek.js]: Moved in da9cadbaf09044c09bd762d03fb08d007a64ea5a

File: src/activityHandler/activityTypes/dailyGoals/doNSeconds.js (hunk 3 above)
Comment (JimenaAndrea) [reply src/activityHandler/activityTypes/dailyGoals/doNSeconds.js]: Moved in da9cadbaf09044c09bd762d03fb08d007a64ea5a

File: src/activityHandler/activityTypes/dailyGoals/doNSeconds.js (hunk 2 above)
Comment (JimenaAndrea) [reply src/activityHandler/activityTypes/dailyGoals/doNSeconds.js]: Moved in da9cadbaf09044c09bd762d03fb08d007a64ea5a

File: App.js (hunk 1 above)
Comment (JimenaAndrea) [reply App.js]: If shouldShowAlert: false the notifications alert doesn't appear with the app open. The notifications code was moved in da9cadbaf09044c09bd762d03fb08d007a64ea5a

File: src/screens/ActivityDetailScreen/TodayPannel.js (hunk 7 above)
Comment (OliverLSanz) [reply src/screens/ActivityDetailScreen/TodayPannel.js]: Done in cba8b240ce63bf962356a45fac887e0419d126cc

**Pull request merged**